### DISCUSSION TITLE
Global whitelist support?

### DISCUSSION THREAD

**Discussion by PeterDaveHello** [0xERR0R/blocky#490] [Q&A]
Not sure if I understand correctly.

Currently, if I have three blacklist groups, name `ads`, `phishing` & `parking-domain`, and I only have one whitelist group, called `safelist`, I want to make sure that all the hosts in the `safelist` will not be blocked, then I can't name the whitelist group `safelist`, instead, I need to duplicate the group three times, and named them `ads`, `phishing` & `parking-domain` accordingly, am I right?

Using a different name for the whitelist will have the result that the document mentioned:

>  If a group has only whitelist entries -> this means only domains from this list are allowed, all other domains will be blocked

Is it possible that we can have a more efficient way to maintain the whitelist group? Or I'm using it the wrong way? Thanks a lot!

**Comment by 0xERR0R** [comment]
This is actually the way how it works. You can define different groups for blocking and a whitelist to "override"/"disable" of some  blocking entries. If you have 3 groups and you want to reuse a whitelist for all three, I think there is currently no way to achieve it without defining 3 whitelists with corresponding names.

**Reply by PeterDaveHello** [reply]
Not sure if the manual duplicate of the groups also makes the memory consumption also become multiple times more? On a small VPS, it seems the memory become not enough once I did it 😅

**Comment by ThinkChaos** [comment]
You should be able to use YAML anchors for this. In this example, b will have the same contents as a when parsed:
```yaml
a: &anchor
  x: y

b:
  <<: *anchor
```

**Comment by PeterDaveHello** [comment]
YAML anchor could be a working approach here, though sometimes not so easy to read or maintain it 😅

**Reply by PeterDaveHello** [reply]
I may remember it wrong, and not familiar enough with YAML, because it's not a key-value pair, it looks like can be done yet.

Refererce:
- https://github.com/yaml/yaml/issues/35
- https://github.com/yaml/yaml/issues/48

**Reply by ThinkChaos** [reply]
Ah that sucks. I didn't test it before posting, sorry!
I don't have this issue because I just use a single whitelist and its empty for now. Also my config is generated (with NixOS) so I could copy sections quite easily.

I think it'd be fair to re-open the issue for this @0xERR0R ?

**Reply by PeterDaveHello** [reply]
No problem at all, I guess my scenario is a little more complex than usual, so that I'll need to have some global/base rules, and then have some additional lists enabled for certain groups.

**Reply by ThinkChaos** [reply]
Just figured out why I thought using anchors would worke: GitLab CI supports it ([docs](https://docs.gitlab.com/ee/ci/yaml/yaml_optimization.html#yaml-anchors-for-scripts)).
The [issue](https://gitlab.com/gitlab-org/gitlab/-/issues/23005) for it says they flatten the array, and indeed if I replace the config section's type in Blocky from `map[string][]string` to `map[string]interface{}`, it unmarshals the following successfully:
```yaml
blackLists:
  default: &test-anchor
  - a
  - b
  other:
  - *test-anchor
```

So it would be possible to handle nested arrays transparently in Blocky to allow using anchors.
Whether that's the best way to achieve this in Blocky is another question though.
